### PR TITLE
[feature] multiple personal websites in user profile

### PR DESCRIPTION
**Pull request by billyhunt**
Fixes:  #2387 

Allow users to add and sort unlimited websites to their personal profile.  

I have a couple of questions.

1.  How many sites should we allow?

2.  What to label the different sites.  Should we add a custom label feature so people can write "Personal," "Job," "School," etc?  Labeling them "Website 1," "Website 2," "Website 3" seems a little sterile.  

Current:

screenshot 2015-04-03 17 47 09

Proposed:

multiple-sites

### CODE REVIEW

**Comment by lbanner** [comment]
Hi @billyhunt -- I think this is in response to a GitHub issue or two, correct? Please note those GH issue numbers in your description/outline above w/ "Fixes # ....." -- Helps for context and to keep up on the issues themselves. Thanks.

File: framework/auth/core.py [hunk 1]
```diff
@@ -81,18 +81,23 @@ def validate_year(item):
...
 validate_url = URLValidator()
-def validate_personal_site(value):
-    if value:
-        try:
-            validate_url(value)
-        except ValidationError:
-            # Reraise with a better message
-            raise ValidationError('Invalid personal URL.')
-
+def validate_profileWebsites(profileWebsites):
```
Comment (sloria) [inline framework/auth/core.py]: Use `snake_case` for function names.

File: website/static/js/profile.js [hunk 2]
```diff
@@ -33,11 +33,25 @@ var cleanByRule = function(rule) {
 
 var noop = function() {};
 
+
```
Comment (sloria) [inline website/static/js/profile.js]: Why the added whitespace?

File: website/static/js/profile.js [hunk 3]
```diff
@@ -33,11 +33,25 @@ var cleanByRule = function(rule) {
...
 var SerializeMixin = function() {};
 
 /** Serialize to a JS Object. */
 SerializeMixin.prototype.serialize = function() {
-    return ko.toJS(this);
+    var serializedData = ko.toJS(this);
+    
+    // Remove blank websites from profileWebsites array before save
+    if (serializedData.urls.crud = "/api/v1/settings/social/") {
```
Comment (sloria) [inline website/static/js/profile.js]: Why is this check necessary?

File: website/static/js/profile.js [hunk 4]
```diff
@@ -33,11 +33,25 @@ var cleanByRule = function(rule) {
...
+        var profileWebsites = serializedData.trackedProperties[0];
```
Comment (sloria) [inline website/static/js/profile.js]: This seems brittle; what if we change the ordering of inputs? Or the order of trackedProperties?

File: website/static/js/profile.js [hunk 5]
```diff
@@ -33,11 +33,25 @@ var cleanByRule = function(rule) {
...
+        if (profileWebsites.length > 1) {
+            serializedData.profileWebsites = $.grep(profileWebsites,function(n){ return(n) });
```
Comment (sloria) [inline website/static/js/profile.js]: Is this for filtering inputs that are truthy? If so, you should be able to use `Array.filter`.

File: website/static/js/profile.js [hunk 6]
```diff
@@ -260,14 +277,21 @@ BaseViewModel.prototype.handleSuccess = function() {
...
 
-BaseViewModel.prototype.handleError = function(response) {
+BaseViewModel.prototype.handleError = function(response) {    
     var defaultMsg = 'Could not update settings';
-    var msg = response.responseJSON.message_long || defaultMsg;
-    this.changeMessage(
-        msg,
-        'text-danger',
+    var msg;
+    if (typeof response.responseJSON != "undefined") {
```
Comment (sloria) [inline website/static/js/profile.js]: As long as you're setting `dataType: json` for all AJAX calls, `response` should be the parsed JSON, and you won't need to access `response.responseJSON`. 

Note: I believe the code that was here before was incorrect.

File: website/static/js/profile.js [hunk 7]
```diff
@@ -463,23 +490,54 @@ var extendLink = function(obs, $parent, label, baseUrl) {
...
+        self, 'Profile Website'
+    );    
+}
+
+
+var SocialViewModel = function(urls, modes) {
+    var self = this;
+    BaseViewModel.call(self, urls, modes);
+    TrackedMixin.call(self);
+    var i;
```
Comment (sloria) [inline website/static/js/profile.js]: Why is this necessary?

File: website/static/js/profile.js [hunk 8]
```diff
@@ -463,23 +490,54 @@ var extendLink = function(obs, $parent, label, baseUrl) {
...
+
+    self.addons = ko.observableArray();
+  
+    self.profileWebsites = ko.observableArray([new ProfileWebsite]); // Initially a single item with a blank first entry
```
Comment (sloria) [inline website/static/js/profile.js]: You shouldn't need an 'empty' item in the observable array

File: website/static/js/profile.js [hunk 9]
```diff
@@ -463,23 +490,54 @@ var extendLink = function(obs, $parent, label, baseUrl) {
...
+    });
 
-    self.addons = ko.observableArray();
 
-    self.personal = extendLink(
+    return obs;
+};
+
+
+function ProfileWebsite() {
```
Comment (sloria) [inline website/static/js/profile.js]: Why is this constructor necessary? Can `profileWebsites` just be an array of 'extended' observables?

**Comment by sloria** [comment]
Initial pass finished. :point_up:

File: website/static/js/profile.js (hunk 6 above)
Comment (billyhunt) [inline website/static/js/profile.js]: I think this might be OK for errors.

File: framework/auth/core.py [hunk 10]
```diff
@@ -84,18 +84,23 @@ def validate_year(item):
...
 validate_url = URLValidator()
-def validate_personal_site(value):
-    if value:
-        try:
-            validate_url(value)
-        except ValidationError:
-            # Reraise with a better message
-            raise ValidationError('Invalid personal URL.')
-
+def validate_profile_websites(profileWebsites):
```
Comment (sloria) [inline framework/auth/core.py]: Use snake_case for variable names.

File: framework/auth/core.py [hunk 11]
```diff
@@ -210,7 +215,7 @@ class User(GuidStoredObject, AddonModelMixin):
         'github': u'http://github.com/{}',
         'scholar': u'http://scholar.google.com/citation?user={}',
         'twitter': u'http://twitter.com/{}',
-        'personal': u'{}',
+        'profileWebsites': u'[{}]',
```
Comment (sloria) [inline framework/auth/core.py]: This doesn't appear to be correct. Should `profileWebsites` be a list of strings?

Comment (sloria) [inline framework/auth/core.py]: @fabianvf If one of these values is a list, will ES queries and results still display correctly?

Comment (sloria) [inline framework/auth/core.py]: Also, will this require a search migration?

File: website/static/js/koHelpers.js [hunk 12]
```diff
@@ -59,6 +59,31 @@ addExtender('trimmed', function(value) {
...
+    if (!value || value.search(/^https?:\/\//i) === 0) {
+        return value;
+    }
+    return 'http://' + value;
+});
+
+addExtender('trimmedWebArray', function(values) {
+    if (values) {
+    for (var i = 0; i < values.length; i++) {   
+        console.log("trimmedhttp value is " + values[i] + "and values.length is " + values.length);
```
Comment (sloria) [inline website/static/js/koHelpers.js]: Don't commit any `console.log` statements.

File: website/static/js/koHelpers.js [hunk 13]
```diff
@@ -59,6 +59,31 @@ addExtender('trimmed', function(value) {
...
+        var value = values[i];
+        var value = $.trim(value);
+        if (!value || value.search(/^https?:\/\//i) === 0) {
```
Comment (sloria) [inline website/static/js/koHelpers.js]: There's duplication here with the `trimmedhttp` extender. This can be refactored.

File: website/static/js/koHelpers.js [hunk 14]
```diff
@@ -59,6 +59,31 @@ addExtender('trimmed', function(value) {
     return $.trim(value);
 });
 
+addExtender('trimmedhttp', function(value) {
```
Comment (sloria) [inline website/static/js/koHelpers.js]: How about `trimmedURL`?

File: website/static/js/koHelpers.js [hunk 15]
```diff
@@ -59,6 +59,31 @@ addExtender('trimmed', function(value) {
...
+ //   console.log("trimmedhttp value is " + value);
+    value = $.trim(value);
+    if (!value || value.search(/^https?:\/\//i) === 0) {
+        return value;
+    }
+    return 'http://' + value;
+});
+
+addExtender('trimmedWebArray', function(values) {
```
Comment (sloria) [inline website/static/js/koHelpers.js]: For consistent naming, how about `trimmedURLArray`?

File: website/static/js/profile.js [hunk 16]
```diff
@@ -38,7 +38,20 @@ var SerializeMixin = function() {};
...
 /** Serialize to a JS Object. */
 SerializeMixin.prototype.serialize = function() {
-    return ko.toJS(this);
+    var serializedData = ko.toJS(this);
+    
+    function removeBlankValues(value) {
+      return value != "";
+    }
+    // Remove blank websites from profileWebsites array before save
+    if (serializedData.urls.crud = "/api/v1/settings/social/") {
```
Comment (sloria) [inline website/static/js/profile.js]: Why not just check for the existence of `serializedData.profileWebsites`?

File: framework/auth/core.py (hunk 11 above)
Comment (fabianvf) [inline framework/auth/core.py]: Should be fine. Any code that accesses/displays it will have to be able to deal with a list, but elasticsearch shouldn't mind. Don't think a migration will be required, but I'll have to look into it.

File: website/static/js/profile.js (hunk 6 above)
Comment (sloria) [inline website/static/js/profile.js]: OK, accessing `responseJSON` is correct.

Consult https://github.com/rwaldron/idiomatic.js#type for info on proper type checking.

```js
response.responseJSON === undefined
```

File: website/static/js/profile.js [hunk 17]
```diff
@@ -468,19 +493,35 @@ var SocialViewModel = function(urls, modes) {
...
     BaseViewModel.call(self, urls, modes);
     TrackedMixin.call(self);
+    var i;
 
     self.addons = ko.observableArray();
+  
+    self.profileWebsites = ko.observableArray(); // Initially a single item with a blank first entry  
+                
+    self.hasProfileWebsites = ko.computed(function() {
+        if (self.profileWebsites())
```
Comment (sloria) [inline website/static/js/profile.js]: Need `{}` for the if block

File: website/static/js/profile.js [hunk 18]
```diff
@@ -468,19 +493,35 @@ var SocialViewModel = function(urls, modes) {
     var self = this;
     BaseViewModel.call(self, urls, modes);
     TrackedMixin.call(self);
+    var i;
 
     self.addons = ko.observableArray();
+  
+    self.profileWebsites = ko.observableArray(); // Initially a single item with a blank first entry
```
Comment (sloria) [inline website/static/js/profile.js]: Why populate this array with a blank item? I think the array can just contain the added entries.

**Comment by sloria** [comment]
Please add new models tests for your changes.

**Comment by sloria** [comment]
Also, what will happen to users who already have a personal website saved in the database? Will this need a database migration?

**Comment by sloria** [comment]
Pass finished. :v:

File: framework/auth/core.py (hunk 11 above)
Comment (sloria) [inline framework/auth/core.py]: Thanks @fabianvf .

@billyhunt Have you checked how this affects the display of search results?

**Comment by GaryKriebel** [comment]
Here is my profile locally before the new code:

screen shot 2015-05-01 at 2 49 01 pm

Here is my profile locally after the new code (Is my website gone?):
screen shot 2015-05-01 at 2 50 22 pm

**Comment by brianjgeiger** [comment]
So that's probably a "yes" on needing a DB migration. :)

**Comment by sloria** [comment]
@billyhunt https://cosdev.readthedocs.org/en/latest/osf/migrations.html

File: framework/auth/core.py (hunk 11 above)
Comment (billyhunt) [inline framework/auth/core.py]: Great question.  Search is not currently searching the new websites.

File: website/static/js/koHelpers.js [hunk 19]
```diff
@@ -44,6 +44,16 @@ var makeRegexValidator = function(regex, message, match) {
...
 };
 
+var cleanURL = function(url) {
+    url = $.trim(url);
+    if (!value || value.search(/^https?:\/\//i) === 0) {
+        return url;
+    }
+    return 'http://' + url;
+
+    }
```
Comment (sloria) [inline website/static/js/koHelpers.js]: Fix indentation of this bracket.

File: website/static/js/koHelpers.js [hunk 20]
```diff
@@ -59,6 +69,22 @@ addExtender('trimmed', function(value) {
...
 });
 
+addExtender('trimmedURL', function(value) {
+    return cleanURL(value);
+
+});
+
+addExtender('trimmedURLArray', function(values) {
+    if (values) {
+    for (var i = 0; i < values.length; i++) {
```
Comment (sloria) [inline website/static/js/koHelpers.js]: Fix indentation.

File: framework/auth/core.py [hunk 21]
```diff
@@ -354,7 +359,7 @@ class User(GuidStoredObject, AddonModelMixin):
     # Social links
     social = fields.DictionaryField(validate=validate_social)
     # Format: {
-    #     'personal': <personal site>,
+    #     'profileWebsites': <profile website>,
```
Comment (sloria) [inline framework/auth/core.py]: Documentation should reflect that `profileWebsites` is a list.

File: website/static/js/profile.js [hunk 22]
```diff
@@ -468,19 +480,35 @@ var SocialViewModel = function(urls, modes) {
...
-            url: true,
-            ensureHttp: true
-        }),
-        self, 'personal'
-    );
+  
+    self.profileWebsites = ko.observableArray();
+                
+    self.hasProfileWebsites = ko.computed(function() {
+        if (self.profileWebsites())
```
Comment (sloria) [inline website/static/js/profile.js]: Use brackets around the `if` block.

File: website/static/js/profile.js [hunk 23]
```diff
@@ -468,19 +480,35 @@ var SocialViewModel = function(urls, modes) {
...
+            }
+        }
+        return false;
+    });
+    
+    self.hasMultiple = ko.computed(function() {
+        if (self.profileWebsites()) {
+            return self.profileWebsites().length > 1;
+        }
+        else return false
```
Comment (sloria) [inline website/static/js/profile.js]: Missing semicolon. Also, use brackets around `else` blocks

File: website/static/js/profile.js [hunk 24]
```diff
@@ -555,8 +609,44 @@ var SocialViewModel = function(urls, modes) {
...
+        serializedData.profileWebsites = serializedData.profileWebsites.filter(removeBlankValues);
+    }     
+    return serializedData;
+};
+
+SocialViewModel.prototype.unserialize = function(data) {
+    var self = this, websiteValue = [];
+    var websiteValue;
+    $.each(data || {}, function(key, value) {
+         if (ko.isObservable(self[key]) && key == "profileWebsites") {
```
Comment (sloria) [inline website/static/js/profile.js]: Use `===` for string comparison.

**Comment by sloria** [comment]
Please use jshint to locate syntax and style errors.

File: website/static/js/profile.js [hunk 25]
```diff
@@ -261,14 +264,21 @@ BaseViewModel.prototype.handleSuccess = function() {
...
-    var msg = response.message_long || defaultMsg;
-    this.changeMessage(
-        msg,
-        'text-danger',
+    var msg;
+    if (response.responseJSON === undefined) {
+        msg = defaultMsg;
+    }
+    else {
+        msg = response.responseJSON.message_long
```
Comment (sloria) [inline website/static/js/profile.js]: Missing semicolon

File: website/static/js/profile.js [hunk 26]
```diff
@@ -522,14 +550,28 @@ var SocialViewModel = function(urls, modes) {
...
 
     var validated = ko.validatedObservable(self);
+ 
     self.isValid = ko.computed(function() {
         return validated.isValid();
     });
     self.hasValidProperty(true);
 
+    self.addWebsite = function(profileWebsite) {
+        var nextItemIndex = self.profileWebsites().length;
```
Comment (sloria) [inline website/static/js/profile.js]: This variable isn't used.

File: website/static/js/profile.js [hunk 27]
```diff
@@ -522,14 +550,28 @@ var SocialViewModel = function(urls, modes) {
...
+        this.profileWebsites.push(ko.observable().extend({
+            trimmed: true,
+            url: true,
+            trimmedhttp: true
+            }));
```
Comment (sloria) [inline website/static/js/profile.js]: Fix indentation

File: website/static/js/profile.js [hunk 28]
```diff
@@ -468,19 +480,35 @@ var SocialViewModel = function(urls, modes) {
...
-            trimmed: true,
-            url: true,
-            ensureHttp: true
-        }),
-        self, 'personal'
-    );
+  
+    self.profileWebsites = ko.observableArray();
+                
+    self.hasProfileWebsites = ko.computed(function() {
```
Comment (sloria) [inline website/static/js/profile.js]: I'm not sure that this is necessary. Only non-empty, valid values should ever be allowed the `profileWebsites()` array. If that's the case, `hasProfileWebsites` is unnecessary because you could just check `if (self.profileWebsites())`.

File: website/static/js/profile.js [hunk 29]
```diff
@@ -468,19 +480,35 @@ var SocialViewModel = function(urls, modes) {
...
+        for (var i=0; i<self.profileWebsites().length; i++) {
+            if (self.profileWebsites()[i]) {
+                return true;
+            }
+        }
+        return false;
+    });
+    
+    self.hasMultiple = ko.computed(function() {
+        if (self.profileWebsites()) {
```
Comment (sloria) [inline website/static/js/profile.js]: This check is unnecessary.

File: website/static/js/profile.js [hunk 30]
```diff
@@ -468,19 +480,35 @@ var SocialViewModel = function(urls, modes) {
...
+            return self.profileWebsites().length > 1;
+        }
+        else return false
+    });
+    
+    self.canRemove = ko.computed(function () {
```
Comment (sloria) [inline website/static/js/profile.js]: This computed is probably unnecessary; it is identical to `hasMultiple`

File: website/static/js/profile.js (hunk 29 above)
Comment (sloria) [inline website/static/js/profile.js]: I'm not even sure this computed is necessary.

File: website/static/js/profile.js [hunk 31]
```diff
@@ -539,9 +581,21 @@ var SocialViewModel = function(urls, modes) {
             {label: 'Google Scholar', text: self.scholar(), value: self.scholar.url()}
         ];
     });
-
+        
+    self.profileWebsiteEmpty = ko.computed(function() {
```
Comment (sloria) [inline website/static/js/profile.js]: If the code is refactored to only allow non-empty, valid values in `profileWebsites` as suggested in my previous comment, this computed is unnecessary.

File: website/static/js/profile.js [hunk 32]
```diff
@@ -555,8 +609,44 @@ var SocialViewModel = function(urls, modes) {
 SocialViewModel.prototype = Object.create(BaseViewModel.prototype);
 $.extend(SocialViewModel.prototype, SerializeMixin.prototype, TrackedMixin.prototype);
 
-var ListViewModel = function(ContentModel, urls, modes) {
+SocialViewModel.prototype.serialize = function() {
+    var serializedData = ko.toJS(this);
+    
+    function removeBlankValues(value) {
```
Comment (sloria) [inline website/static/js/profile.js]: This code can also be removed if the code is refactored to not allow blank values into `profileWebsites`.

File: website/static/js/profile.js [hunk 33]
```diff
@@ -766,13 +856,17 @@ var Jobs = function(selector, urls, modes) {
...
 };
 
+
+
 module.exports = {
     Names: Names,
     Social: Social,
     Jobs: Jobs,
     Schools: Schools,
+ //   PersonalWebsites: PersonalWebsites,
```
Comment (sloria) [inline website/static/js/profile.js]: Remove unnecessary comments.

File: website/templates/include/profile/social.mako [hunk 34]
```diff
@@ -1,83 +1,119 @@
...
     <div data-bind="if: mode() === 'edit'">
-
         <form role="form" data-bind="submit: submit">
 
-            <div class="form-group">
-                <label>Personal Site</label>
-                <div class="input-group">
-                <span class="input-group-addon"><i class="glyphicon glyphicon-globe"></i></span>
-                <input class="form-control" data-bind="value: personal" placeholder="http://personalsite.com"/>
+    <div data-bind="sortable: {
```
Comment (sloria) [inline website/templates/include/profile/social.mako]: Fix indentation.

**Comment by sloria** [comment]
Pass finished. This will also need a migration script to update the `social` field.

File: website/static/js/profile.js (hunk 31 above)
Comment (billyhunt) [inline website/static/js/profile.js]: I use profileWebsiteEmpty to tell if the add another button should be in the view.  If there is a blank website, then you should not need to add another one.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 35]
```diff
@@ -0,0 +1,54 @@
...
+logger = logging.getLogger(__name__)
+
+
+#def main(dry=True):
+#    init_app(set_backends=True, routes=False, mfr=False)  # Sets the storage backends on all models
+#    do_migration(get_targets(), dry=dry)
+
+    
+def main():
+    print "in main"
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: Remove debugging statements.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 36]
```diff
@@ -0,0 +1,54 @@
...
+            if user.social.get('personal'):
+                migrate_personal_to_profile_websites(user)
+        logger.info(repr(user.social))
+        if not dry_run:
+            user.save()
+
+
+def get_users_with_personal_websites():
+    return models.User.find(
+        Q('social', 'ne', None)
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: I believe you can do a query on nested keys, e.g. `Q('social.personal..)`

Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: Make sure to test this out, though.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 37]
```diff
@@ -0,0 +1,54 @@
...
+    dry_run = 'dry' in sys.argv
+    if not dry_run:
+        scripts_utils.add_file_logger(logger, __file__)
+    logger.info("migrating personal to profileWebsites")
+            
+    for user in get_users_with_personal_websites():
+        logger.info(repr(user))
+        logger.info(repr(user.social))
+        if not user.social.get('profileWebsites'):
+            user.social['profileWebsites'] = [u'']
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: Why include the empty string? If a user does not have any personal websites, the list should be empty.

File: website/static/js/koHelpers.js [hunk 38]
```diff
@@ -59,6 +69,22 @@ addExtender('trimmed', function(value) {
...
+addExtender('trimmedURL', function(value) {
+    return cleanURL(value);
+
+});
+
+addExtender('trimmedURLArray', function(values) {
+    if (values) {
+        for (var i = 0; i < values.length; i++) {   
+            values[i] = cleanURL(values[i]);
+            }
```
Comment (sloria) [inline website/static/js/koHelpers.js]: Fix indentation.

File: website/static/js/profile.js [hunk 39]
```diff
@@ -468,19 +479,37 @@ var SocialViewModel = function(urls, modes) {
...
+    self.canEditWebsite = ko.observable(false);
+
+    self.editWebsiteButton =  function(){
+        self.canEditWebsite(!self.canEditWebsite());
+    }
+
+    self.profileWebsites = ko.observableArray();
+                
+    self.hasProfileWebsites = ko.computed(function() {
+        if (self.profileWebsites())
```
Comment (sloria) [inline website/static/js/profile.js]: Syntax error: a `{` is missing here.

Please use jshint to catch syntax and style errors.

File: website/static/js/profile.js [hunk 40]
```diff
@@ -468,19 +479,37 @@ var SocialViewModel = function(urls, modes) {
...
+            if (self.profileWebsites()[i]) {
+                return true;
+            }
+        }
+        return false;
+    });
+    
+    self.canRemove = ko.computed(function () {
+        if (self.profileWebsites()) {
+            return self.profileWebsites().length > 1;
```
Comment (sloria) [inline website/static/js/profile.js]: Why should a user be required to have at least 1 website? Why can't they have 0?

File: website/static/js/profile.js (hunk 31 above)
Comment (sloria) [inline website/static/js/profile.js]: There should never be any blank websites. A user's `social.profileWebsites` field should only contain non-empty, valid values.

File: website/static/js/profile.js [hunk 41]
```diff
@@ -539,9 +568,21 @@ var SocialViewModel = function(urls, modes) {
             {label: 'Google Scholar', text: self.scholar(), value: self.scholar.url()}
         ];
     });
-
+        
+    self.profileWebsiteEmpty = ko.computed(function() {
+        if (self.profileWebsites())
```
Comment (sloria) [inline website/static/js/profile.js]: Missing `{}` around `if` block.

File: website/static/js/profile.js [hunk 42]
```diff
@@ -550,13 +591,65 @@ var SocialViewModel = function(urls, modes) {
...
     });
 
+    self.addWebsite = function() {
+        this.profileWebsites.push(ko.observable().extend({
+            trimmedURL: true
+            }));
+    }
+    
+    self.removeWebsite = function(profileWebsite) {
+        for (var i=0; i < self.profileWebsites().length; i++) {
```
Comment (sloria) [inline website/static/js/profile.js]: You should be able to use `Array.indexOf` instead of a for loop.

File: website/static/js/profile.js [hunk 43]
```diff
@@ -550,13 +591,65 @@ var SocialViewModel = function(urls, modes) {
...
+            if (profileWebsite == ko.toJS(self.profileWebsites()[i]))
+                self.profileWebsites.splice(i, 1);
+        }
+    }
+
+
+    if (self.profileWebsites().length == 0) {
```
Comment (sloria) [inline website/static/js/profile.js]: Use `===` for comparison with numbers (jshint).

**Comment by sloria** [comment]
Clicking to remove a single website removes two of the rows; check your logic there.

pr

File: website/templates/include/profile/social.mako [hunk 44]
```diff
@@ -1,86 +1,106 @@
...
+                            handle: '.sort-handle',
+                            containment: '#containDrag'
+                        }
+                    }">
+
+                <div>
+                    <div class="sort-handle">
+                        <i title="Click to Remove" class="btn text-danger pull-right  fa fa-times fa" data-bind="click: $parent.removeWebsite"></i>
+                        <div class="input-group" >
+                            <span class="input-group-addon"><i title="Drag to Reorder"  class="glyphicon glyphicon-menu-hamburger"></i></span>
```
Comment (sloria) [inline website/templates/include/profile/social.mako]: Prefer fontawesome to glyphicons. This might be what you need here: https://fortawesome.github.io/Font-Awesome/icon/bars/

File: scripts/migration/migrate_personal_to_profile_websites.py (hunk 37 above)
Comment (billyhunt) [inline scripts/migration/migrate_personal_to_profile_websites.py]: If the list is empty, and the list creation fails on the client side, then the user will not see a website field to fill out for their initial profile.  This happened with one user with intentionally corrupted data in my local environment, and wanted to be safe.  There might be a more elegant/efficient way to to handle it though.

File: framework/auth/core.py [hunk 45]
```diff
@@ -830,12 +833,15 @@ def is_confirmed(self):
...
-        return {
-            key: self.SOCIAL_FIELDS[key].format(val)
-            for key, val in self.social.items()
-            if val and
-            self.SOCIAL_FIELDS.get(key)
-        }
+        social_user_fields = {}
+        #import ipdb;ipdb.set_trace()
+        for key, val in self.social.items():
+            if val and key in self.SOCIAL_FIELDS.keys():
```
Comment (sloria) [inline framework/auth/core.py]: `.keys()` is unnecessary, `...key in self.SOCIAL_FIELDS`

File: framework/auth/core.py [hunk 46]
```diff
@@ -830,12 +833,15 @@ def is_confirmed(self):
...
     @property
     def social_links(self):
-        return {
-            key: self.SOCIAL_FIELDS[key].format(val)
-            for key, val in self.social.items()
-            if val and
-            self.SOCIAL_FIELDS.get(key)
-        }
+        social_user_fields = {}
+        #import ipdb;ipdb.set_trace()
```
Comment (sloria) [inline framework/auth/core.py]: Remove

File: website/static/js/profile.js [hunk 47]
```diff
@@ -550,13 +591,65 @@ var SocialViewModel = function(urls, modes) {
...
 };
 SocialViewModel.prototype = Object.create(BaseViewModel.prototype);
 $.extend(SocialViewModel.prototype, SerializeMixin.prototype, TrackedMixin.prototype);
 
-var ListViewModel = function(ContentModel, urls, modes) {
+SocialViewModel.prototype.serialize = function() {
+    var serializedData = ko.toJS(this);
+    var profileWebsites = serializedData.profileWebsites;
+    
+    function removeBlankValues(value) {
```
Comment (sloria) [inline website/static/js/profile.js]: Again, if we don't allow blank values to be added to the list of website, this becomes unnecessary.

File: website/static/js/profile.js [hunk 48]
```diff
@@ -468,19 +479,37 @@ var SocialViewModel = function(urls, modes) {
...
-    );
+    self.canEditWebsite = ko.observable(false);
+
+    self.editWebsiteButton =  function(){
+        self.canEditWebsite(!self.canEditWebsite());
+    }
+
+    self.profileWebsites = ko.observableArray();
+                
+    self.hasProfileWebsites = ko.computed(function() {
```
Comment (sloria) [inline website/static/js/profile.js]: This, too, becomes unnecessary if blank values are disallowed because you would just be able to check `if (self.profileWebsites())`

File: website/static/js/profile.js [hunk 49]
```diff
@@ -468,19 +479,37 @@ var SocialViewModel = function(urls, modes) {
...
-        ko.observable().extend({
-            trimmed: true,
-            url: true,
-            ensureHttp: true
-        }),
-        self, 'personal'
-    );
+    self.canEditWebsite = ko.observable(false);
+
+    self.editWebsiteButton =  function(){
```
Comment (sloria) [inline website/static/js/profile.js]: This doesn't seem to be used anywhere. Same with `canEditWebsite`. Remove them if not needed.

**Comment by sloria** [comment]
Pass finished. :facepunch:

File: scripts/migration/migrate_personal_to_profile_websites.py (hunk 36 above)
Comment (billyhunt) [inline scripts/migration/migrate_personal_to_profile_websites.py]: I tried it initially, and it did not work.  I will double check.

Comment (billyhunt) [inline scripts/migration/migrate_personal_to_profile_websites.py]: I can do nested queries.  Would you prefer me to do 2 queries for the objects I want (personal and profileWebsites), or just run the one query to get the whole social object?

File: scripts/migration/migrate_personal_to_profile_websites.py (hunk 37 above)
Comment (billyhunt) [inline scripts/migration/migrate_personal_to_profile_websites.py]: I fixed this making sure the data is taken care of on the client side.

File: website/static/js/profile.js (hunk 40 above)
Comment (billyhunt) [inline website/static/js/profile.js]: canRemove is used to tell if we should remove the dialog box when you hit the remove button, or just clear the field.  So when you hit remove on the last website, it does not look like this.  

image

The blank sites are stripped before uploading to the server.

File: website/static/js/profile.js (hunk 47 above)
Comment (billyhunt) [inline website/static/js/profile.js]: The profileWebsites observableArray currently needs an entry to display a blank field for the user to edit.  I changed this to add a single entry on the client side if the server list is empty, and to strip all values before sending to the server as opposed to sending a single blank item in the list.

File: website/static/js/profile.js [hunk 50]
```diff
@@ -550,13 +577,68 @@ var SocialViewModel = function(urls, modes) {
         return false;
     });
 
+    self.addWebsite = function() {
```
Comment (sloria) [inline website/static/js/profile.js]: This name could be improved; `addWebsite` implies that the method is used to add a new website to a user's profile; in reality, this adds a blank input. Maybe `addWebsiteInput`?

**Comment by sloria** [comment]
https:// URLs are not being properly handled. I get `Invalid personal URL` when I input a valid https URL

File: website/static/js/profile.js [hunk 51]
```diff
@@ -550,13 +577,68 @@ var SocialViewModel = function(urls, modes) {
...
+        this.profileWebsites.push(ko.observable().extend({
+            trimmedURL: true
+            }));
```
Comment (sloria) [inline website/static/js/profile.js]: Fix indentation.

File: scripts/migration/migrate_personal_to_profile_websites.py (hunk 36 above)
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: The query should be constructed such that the script can be run more than once. If it is run more than once, a user should not have the same personal website appear twice in `profileWebsite`. This can currently happen, and is therefore a bug.

**Comment by sloria** [comment]
Pass finished. :rice_ball:

**Comment by sloria** [comment]
Putting this on hold for the time being until the end of this epic.

**Comment by lyndsysimon** [comment]
Removing this from "hold" status as the epic has ended.

**Comment by sloria** [comment]
Ping @lbanner @GaryKriebel -- is this still a desired feature? Close, hold, or continue development?

**Comment by GaryKriebel** [comment]
I would think there would be value to the user to have more than one website listed.... I would vote for continue.

**Comment by lbanner** [comment]
OK -- yes, continue.

**Comment by sloria** [comment]
Ping @erinmayhood for review of this feature before we do further development on this.

**Comment by billyhunt** [comment]
Replaced by https://github.com/CenterForOpenScience/osf.io/pull/4263
